src/simulator/services/persistent_state.py

```diff
@@ -0,0 +1,184 @@
+
+    def _add_colour(self, name: str, colour: Colour, visible: bool) -> DynamicColour:
+        if name in self.colours:
+            return self.colours[name]
+        dc = self.colours[name] = DynamicColour(colour, name, self.__colour_callback, visible)
+        self.__colour_callback(dc)
+        return dc
+
+    def _from_view(self, other: 'PersistentStateView') -> None:
+        self.colours = {k: v for k, v in self.colours.items() if k in other.colours}
+        for n, c in other.colours.items():
+            if n not in self.colours:
+                self._add_colour(n, c.colour, c.visible)
+            else:
+                self.colours[n].set_all(c.colour, c.visible)
+
+    def _from_json(self, data: Dict[str, Any]) -> None:
+        for n, c in data["colours"].items():
+            colour = Colour(*c["colour"])
+            visible = bool(c["visible"])
+            self.colours[n] = DynamicColour(colour, n, self.__colour_callback, visible)
+
+    def _to_json(self) -> Dict[str, Any]:
+        data = {}
+        dc = data["colours"] = {}
+        for n, c in self.colours.items():
+            dc[n] = {}
+            dc[n]["colour"] = tuple(c.colour)
+            dc[n]["visible"] = c.visible
+        return data
+
+    def is_effective(self) -> bool:
+        return self.index == PersistentStateView.EFFECTIVE_VIEW
+
+class PersistentState(Service):
+    file_name: str
+
+    MAX_VIEWS: Final[int] = 6
+    effective_view: PersistentStateView
+    views: List[PersistentStateView]
+    __view_idx: int
+
+    _save_task: Optional['Task']
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **exclude_from_dict(kwargs, ["file_name"]))
+        self.file_name = kwargs["file_name"] if "file_name" in kwargs else ".pathbench.json"
+
+        self._save_task = None
+
+        self.reset(False)
+        self.load()
+        self.save()
+
+    def reset(self, save: bool = True) -> None:
+        self.effective_view = PersistentStateView(self._services, self, PersistentStateView.EFFECTIVE_VIEW)
+        self.views = [PersistentStateView(self._services, self, i) for i in range(self.MAX_VIEWS)]
+        self.__view_idx = 0
+        if save:
+            self.save()
+
+    def load(self) -> None:
+        def check(req: bool) -> None:
+            if not req:
+                raise RuntimeError
+
+        if os.path.isfile(self.file_name):
+            self._services.debug.write("Loading state from '{}'".format(self.file_name), DebugLevel.BASIC)
+            with open(self.file_name, 'r') as f:
+                try:
+                    jdata = json.load(f)
+                    jidx = jdata["view_index"]
+                    check(jidx >= 0 and jidx < self.MAX_VIEWS)
+                    jviews = jdata["views"]
+                    for v in range(self.MAX_VIEWS):
+                        self.views[v]._from_json(jviews[v])
+                    self.view_idx = jidx
+                except Exception:
```

Comment:
This isn't great; you should only catch certain exceptions. Also, since `check` is only used once, can you replace that call with an inline `if ... raise RuntimeError("msg")`?

Reply:
-- to clarify - don't catch `Exception`, just catch the specific exceptions you expect to be thrown, for example maybe `RuntimeError` here.

Reply:
Moved index checking to `@view_idx.setter`.

Reply:
Added outputting details of exception, so unexpected exceptions won't go unnoticed.